### PR TITLE
Updated perf_ ntttcp_client.sh : Support for UDP on ntttcp

### PR DESCRIPTION
**Pull request by asofron**

### CODE REVIEW

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 1]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
     rm -rf $HOME/$log_folder
 fi
 
-mkdir $HOME/$log_folder
-# Determine test type lagscope/ntttcp
-if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp" ]]; then
+mkdir  "$HOME/$log_folder"
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: 2 spaces, leave it one

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 2]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+# Determine test type lagscope/ntttcp/ntttcp-udp
+if [[ "${UDP_MODE}" = "yes" ]]; then
     eth_log="$HOME/$log_folder/eth_report.log"
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: you can move this at the beggining of the script

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 3]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+    "cat /proc/interrupts > $HOME/$log_folder/receiver-interrupts-start.log"
+    previous_tx_bytes="$(get_tx_bytes "$ETH_NAME")"
+    previous_tx_pkts="$(get_tx_pkts "$ETH_NAME")"
     port_change=false
     port=
     for current_test_threads in "${TEST_THREADS[@]}"
     do
-        if [ $current_test_threads -lt $MAX_THREADS ]
+        if [ "$current_test_threads" -lt "$MAX_THREADS" ]
+        then
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: can you use one type of 'if statement'? I mean in some locations is 'if cmd ;; then' and in other is
'if cmd
then'

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 4]
```diff
@@ -581,28 +697,36 @@ if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp"
...
+    cat /proc/interrupts > "$HOME/$log_folder/sender-interrupts-end.log"
+    ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+     "cat /proc/interrupts > $HOME/$log_folder/receiver-interrupts-end.log"
     LogMsg "Ntttcp succeeded with all connections."
     echo "Ntttcp succeeded with all connections." >> ~/summary.log
-    cd $HOME
-    scp -i $HOME/.ssh/${SSH_PRIVATE_KEY} -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@[${STATIC_IP2}]:/root/$log_folder/* /root/$log_folder
-    if [ $? -ne 0 ]; then
+    cd $HOME || exit
+   if !  scp -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@[${STATIC_IP2}]:"/root/$log_folder/* /root/$log_folder"; then
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: 2 spaces between ! and scp

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 5]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+         "pkill -x mpstat"
+    fi
+    ethtool -S eth1 > "$HOME/$log_folder/sender-ethtool-eth1.log"
+    ethtool -S enP2p0s2 > "$HOME/$log_folder/sender-ethtool-enP2p0s2.log"
+    ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+     "ethtool -S eth1 > $HOME/$log_folder/receiver-ethtool-eth1.log"
+    ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+     "ethtool -S enP2p0s2 > $HOME/$log_folder/receiver-ethtool-enP2p0s2.log"
+    cat /proc/interrupts > "$HOME/$log_folder/sender-interrupts-end.log"
+    ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2}  \
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: 2 spaces between {STATIC_IP2} and  \

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 6]
```diff
@@ -544,34 +651,43 @@ if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp"
...
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "mpstat -P ALL 1 ${TEST_DURATION} > $HOME/$log_folder/mpstat-receiver-p${current_test_threads}.log"
+
+        sar -n DEV 1 "${TEST_DURATION}" > "$HOME/$log_folder/sar-sender-p${current_test_threads}.log" &
+        dstat -dam 1 "${TEST_DURATION}" > "$HOME/$log_folder/dstat-sender-p${current_test_threads}.log" &
+        mpstat -P ALL 1 "${TEST_DURATION}" > "$HOME/$log_folder/mpstat-sender-p${current_test_threads}.log" &
+
+        lagscope -s"${SERVER_IP}" -t "${TEST_DURATION}" -V "${ipVersion}" > \
+         "./$log_folder/lagscope-ntttcp-p${num_threads_P}X${num_threads_n}.log" &
+        ntttcp -s "${SERVER_IP}" -P "${num_threads_P}" -n "${num_threads_n}" -t "${TEST_DURATION}" "${ipVersion}"  -f"${port}" > \
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: 2 spaces between "${ipVersion}" and -f"${port}"

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 7]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+        else
+            port=40000
+            port_change=true
+        fi
+        echo "======================================"
+        echo "Running Test: $num_threads_P X $num_threads_n"
+        echo "======================================"
+        ntttcp_recv_log="$HOME/$log_folder/ntttcp-receiver-p${num_threads_P}X${num_threads_n}.log"
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "ulimit -n 20480 && ntttcp -u -b ${NTTTCP_BUFFER} -r ${SERVER_IP} -P ${num_threads_P}  -e ${ipVersion} > ${ntttcp_recv_log}"
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: 2 spaces between ${num_threads_P} and -e

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 8]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "dstat -dam 1 ${TEST_DURATION} > $HOME/$log_folder/dstat-receiver-p${current_test_threads}.log"
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "mpstat -P ALL 1 ${TEST_DURATION} > $HOME/$log_folder/mpstat-receiver-p${current_test_threads}.log"
+
+        sar -n DEV 1 "${TEST_DURATION}" > "$HOME/$log_folder/sar-sender-p${current_test_threads}.log" &
+        dstat -dam 1 "${TEST_DURATION}" > "$HOME/$log_folder/dstat-sender-p${current_test_threads}.log" &
+        mpstat -P ALL 1 "${TEST_DURATION}" > "$HOME/$log_folder/mpstat-sender-p${current_test_threads}.log" &
+
+        lagscope -s "${SERVER_IP}" -t "${TEST_DURATION}" -V  "${ipVersion}" > "./$log_folder/lagscope-ntttcp-p${num_threads_P}X${num_threads_n}.log" &
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: 2 spaces between -V and "${ipVersion}"

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 9]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+    LogMsg "Ntttcp UDP succeeded with all connections."
+    echo "Ntttcp  UDP succeeded with all connections." >> ~/summary.log
+    cd $HOME || exit
+   if ! scp -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no \
+    ${SERVER_OS_USERNAME}@[${STATIC_IP2}]:/root/"$log_folder"/* /root/"$log_folder"; then
+        echo "ERROR: Unable to transfer server side logs."
+        UpdateTestState $ICA_TESTFAILED
+    fi
+if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp" ]]; then
+    eth_log="$HOME/$log_folder/eth_report.log"
```
Comment (adriansuhov) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: as you can see, this is duplicated from above, and that's why it's better to be put at the top

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 10]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
     rm -rf $HOME/$log_folder
 fi
 
-mkdir $HOME/$log_folder
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: why are you removing the if?

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 11]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
-# Determine test type lagscope/ntttcp
-if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp" ]]; then
+mkdir  "$HOME/$log_folder"
+# Determine test type lagscope/ntttcp/ntttcp-udp
+if [[ "${UDP_MODE}" = "yes" ]]; then
     eth_log="$HOME/$log_folder/eth_report.log"
-    echo "#test_connections    throughput_gbps    average_packet_size" > $eth_log
+    echo "#test_connections    throughput_gbps    average_packet_size" > "$eth_log"
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: never use "$eth_log", always use "${eth_log}". this applies for all strings with vars

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 12]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
     sleep 10
     #Starting test
-    cat /proc/interrupts > $HOME/$log_folder/sender-interrupts-start.log
-    ssh -i $HOME/.ssh/${SSH_PRIVATE_KEY} -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} "cat /proc/interrupts > $HOME/$log_folder/receiver-interrupts-start.log"
-    previous_tx_bytes=$(get_tx_bytes $ETH_NAME)
-    previous_tx_pkts=$(get_tx_pkts $ETH_NAME)
+    cat /proc/interrupts > "$HOME/$log_folder/sender-interrupts-start.log"
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: always: strings should be inside quotes

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 13]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+    ssh -i "${HOME}/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+    "cat /proc/interrupts > $HOME/$log_folder/receiver-interrupts-start.log"
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: always leave at least a 4 spaces indentation for blocks

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 14]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+        if [ "$current_test_threads" -lt "$MAX_THREADS" ]
+        then
+            num_threads_P=$current_test_threads
+            num_threads_n=1
+        else
+            num_threads_P=$MAX_THREADS
+            num_threads_n=$(("$current_test_threads" / "$num_threads_P"))
+        fi
+        done
+        if ${port_change}
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: ditto

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh (hunk 7 above)
Comment (ader1990) [reply WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: line too long

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 15]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+        fi
+        echo "======================================"
+        echo "Running Test: $num_threads_P X $num_threads_n"
+        echo "======================================"
+        ntttcp_recv_log="$HOME/$log_folder/ntttcp-receiver-p${num_threads_P}X${num_threads_n}.log"
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "ulimit -n 20480 && ntttcp -u -b ${NTTTCP_BUFFER} -r ${SERVER_IP} -P ${num_threads_P}  -e ${ipVersion} > ${ntttcp_recv_log}"
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "lagscope -r${SERVER_IP} ${ipVersion}"
+         tcp_con_logs="$HOME/$log_folder/tcp-connections-p${current_test_threads}.log"
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: indentation

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh (hunk 8 above)
Comment (ader1990) [reply WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: line too long

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 16]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+        Throughput=$(echo "scale=2;$bytes_new/$TEST_DURATION*8/1024/1024/1024" | bc)
+        previous_tx_bytes=$current_tx_bytes
+        previous_tx_pkts=$current_tx_pkts
+
+        echo "Throughput (gbps): $Throughput"
+        echo "average packet size: $avg_pkt_size"
+        printf "%4s  %8.2f  %8.2f\\n" "${current_test_threads}" "${Throughput}" "${avg_pkt_size}" >> "${eth_log}"
+
+        echo "current test finished. wait 10 seconds for next one... "
+        sleep 10
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: why is the sleep here?

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 17]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+        pkill -x ntttcp
+        pkill -x lagscope
+        pkill -f ESTA
+        pkill -x sar
+        pkill -x dstat
+        pkill -x mpstat
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: the pkill thing is a very dangerous way of killing process, without any checks or traceability

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 18]
```diff
@@ -510,28 +510,135 @@ if [ -d  $HOME/$log_folder ]; then
...
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "pkill -x lagscope"
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "pkill -x sar"
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "pkill -x dstat"
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "pkill -x mpstat"
+    fi
+    ethtool -S eth1 > "$HOME/$log_folder/sender-ethtool-eth1.log"
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: who is eth1 or enP2p0s2 ? this is a hardcoded string and should not be here. use the variables from test params

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 19]
```diff
@@ -612,9 +736,7 @@ else
     echo "ERROR: Unknown test type ${TEST_TYPE}."
     LogMsg "ERROR: Unknown test type ${TEST_TYPE}."
 fi
-zip -r $log_folder.zip $log_folder/*
-
-if [ $sts -eq 0 ]; then
+if ! zip -r "$log_folder.zip" "$log_folder/*"; then
     LogMsg "Test completed"
     echo "Test completed" >> ~/summary.log
 else
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: it seems to me that you have duplicated code, this patch needs refactorization. never duplicate code, use functions if you have common code

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 20]
```diff
@@ -544,7 +544,16 @@ if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp"
         echo "======================================"
         echo "Running Test: $num_threads_P X $num_threads_n"
         echo "======================================"
-        ssh -i $HOME/.ssh/${SSH_PRIVATE_KEY} -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} "ulimit -n 20480 && ntttcp -r${SERVER_IP} -P $num_threads_P -e ${ipVersion} > $HOME/$log_folder/ntttcp-receiver-p${num_threads_P}X${num_threads_n}.log"
+        if [[ "${UDP_MODE}" = "yes" ]]; 
+        # runs ntttcp on udp if the variable exists in test parameters
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: put then on the same if line

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 21]
```diff
@@ -544,7 +544,16 @@ if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp"
...
+        then
+        ntttcp_recv_log="$HOME/$log_folder/ntttcp-receiver-p${num_threads_P}X${num_threads_n}.log"
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: this should be indented

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 22]
```diff
@@ -544,7 +544,16 @@ if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp"
...
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
+         "ulimit -n 20480 && ntttcp -u -b ${NTTTCP_BUFFER} -r ${SERVER_IP} -P ${num_threads_P}  -e ${ipVersion} > ${ntttcp_recv_log}"
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: ditto

File: WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh [hunk 23]
```diff
@@ -544,7 +544,16 @@ if [[ "${TEST_TYPE:-UNDEFINED}" = "UNDEFINED" ]] || [[ "${TEST_TYPE}" = "ntttcp"
...
+        else
+        ssh -i "$HOME/.ssh/${SSH_PRIVATE_KEY}" -f -o StrictHostKeyChecking=no ${SERVER_OS_USERNAME}@${STATIC_IP2} \
```
Comment (ader1990) [inline WS2012R2/lisa/remote-scripts/ica/perf_ntttcp_client.sh]: ditto

**Pull request merged**